streamer/pulp/streamer/server.py

```diff
@@ -378,3 +376,60 @@ def write(self, data):
         :type  data: str
         """
         reactor.callFromThread(self.request.write, data)
+
+
+class SessionCache(Cache):
+    """
+    Session cache.
+    Extends generic cache to derive the key using the URL and
+    downloader configuration.
+    """
+
+    @staticmethod
+    def key(url, downloader):
+        """
+        Get a cache key for the downloader.
+
+        :param url: The download URL.
+        :type url: basestring
+        :param downloader: A file downloader.
+        :type downloader: nectar.downloaders.base.Downloader
+        :return: A cache key based on downloader configuration.
+        """
+        url = urlparse(url)
+        properties = (
+            url.hostname,
+            url.port,
+            downloader.config.basic_auth_username,
+            downloader.config.basic_auth_password,
+            downloader.config.proxy_url,
+            downloader.config.proxy_port,
+            downloader.config.proxy_username,
+            downloader.config.proxy_password,
+            downloader.config.ssl_client_cert
+        )
+        return hash(properties)
+
+    def get(self, url, downloader):
```

Comment:
What do you think about renaming this to something like ``get_or_create()``? That would make it obvious that the behavior is different from the super class ``get()`` method, and also makes me feel better about the function signatures being different (both in parameters and "raises"). :smile:

I also think that intuitively, anyone working with a cache object may reasonably assume that a ``get()`` method will sometimes miss, and expect they need to handle that. I like the idea of using the name to communicate that this method never misses!